### ISSUE TITLE
use DOMContentLoaded instead of load for scopePollingDetectionStrategy

### ISSUE THREAD

**Issue by ArpitAgrawal321** [solana-labs/wallet-adapter#317]
Currently the detection mechanism waits for the `load` event to check for the wallets. This can take a lot of time incase a resource such as an image required by the page keeps loading for a while. This prevents the user from getting the correct UI, does not portray correct state of installed/loadable wallets.
We can use `DOMContentLoaded` instead of `load`  to `performDetection` once the DOM structure is loaded.

Current implementation:
```
function scopePollingDetectionStrategy(performDetection: () => boolean): void {
    if (typeof window === 'undefined' || typeof document === 'undefined') return;
    if (document.readyState === 'complete') {
        performDetection();
        return;
    }
    function listener() {
        window.removeEventListener('load', listener);
        performDetection();
    }
    window.addEventListener('load', listener);
}
```

Suggested Change:
```
function scopePollingDetectionStrategy(performDetection: () => boolean): void {
    if (typeof window === 'undefined' || typeof document === 'undefined') return;
    if (document.readyState === 'complete') {
        performDetection();
        return;
    }
    function listener() {
        window.removeEventListener('DOMContentLoaded', listener);
        performDetection();
    }
    window.addEventListener('DOMContentLoaded', listener);
}
```

**Comment by ArpitAgrawal321** [comment]
Found #215 already exists, closing.

**Comment by jordansexton** [comment]
#215 is just a proposal -- most of the work is on wallets to implement. If that's the only reason to close this, I'd suggest reopening and let's see if it makes sense to PR.

**Comment by jordansexton** [comment]
Related: https://github.com/solana-labs/wallet-adapter/pull/327

@sjchmiela + @vpontis wdyt?

**Comment by steveluscher** [comment]
Great suggestion, @ArpitAgrawal321! I'm trying to load this all back into my head to figure out why I didn't do this in the first place.

I _think_ maybe it's because the check for `document.readyState` was equivalent to listening to `DOMContentLoaded` in most cases, because usually the wallet code _by definition_ doesn't execute until after `DOMContentLoaded` has fired. This might not be true for all apps, so this could very well be worth adding.

I dug around and found some of my early exploration into this with Phantom: https://jsfiddle.net/tb8gy7fc/1/

In any case, we should definitely:

1. Do what you've suggested.
2. Only set up the `DOMContentLoaded` trigger if the `document.readyState === 'loading'` (ie. otherwise we're past the point of `DOMContentLoaded` having fired already)
3. _Keep_ the `load` trigger, in case the wallet still isn't ready on `DOMContentLoaded`.

**Comment by safaiyeh** [comment]
Hey from Magic Eden - we did a patch package to use DOMContentLoaded instead of load since our pages are image dependent. Happy to implement @steveluscher's suggestion to push out the optimization

**Comment by jordansexton** [comment]
@safaiyeh nice! And yes please, that would be great!

**Comment by safaiyeh** [comment]
cool opened here! https://github.com/solana-labs/wallet-adapter/pull/346

**Comment by steveluscher** [comment]
Closed by #346!
